### DISCUSSION TITLE
디미터 법칙이 무엇이고, 이에 대한 장단점을 설명해주세요.

### DISCUSSION THREAD

**Discussion by gyungchan-jo** [2024-woowacourse-study/level-interview#109] [OOP]
.

**Comment by gyungchan-jo** [comment]
디미터 법칙(Law of Demeter 또는 LoD)은 객체 지향 설계의 한 원칙으로, 객체가 다른 객체와 통신할 때 객체는 자신이 가지고 있는 직접적인 관계만을 사용해야 한다는 것을 의미한다. 즉, 객체는 자신의 멤버 변수, 메서드의 매개변수, 메서드 내에서 생성된 객체만을 호출해야 하며, 객체 내부의 객체에 대한 접근은 최소화해야 한다.

장점: 
1. 결합도 감소: 디미터 법칙을 준수하면 객체 간의 결합도가 감소합니다. 객체가 직접적인 관계만을 사용하기 때문에 객체 간의 의존성이 최소화되며, 변경에 대한 영향 범위가 줄어듭니다.
2. 유지보수성 향상: 결합도의 감소로 인해 코드의 유지보수성이 향상됩니다. 하나의 객체를 수정할 때 다른 객체에 대한 변경 사항을 최소화할 수 있습니다.
3. 코드의 가독성 향상: 디미터 법칙을 따르면 객체 간의 관계가 명확하게 드러나기 때문에 코드가 더욱 직관적이고 이해하기 쉽게 됩니다.

단점:
1. 간접적인 호출의 증가: 디미터 법칙을 따르면 간접적인 호출이 많아질 수 있습니다. 이는 코드의 길이를 증가시킬 수 있습니다.
2. 코드의 복잡성 증가: 디미터 법칙을 엄격하게 적용하면 코드의 복잡성이 증가할 수 있습니다. 너무 많은 중간 객체를 추가할 경우 코드가 복잡해지고 이해하기 어려워질 수 있습니다.
3. 성능 저하: 간접적인 호출로 인해 성능 저하가 발생할 수 있습니다. 간접적으로 호출된 메서드의 호출 비용이 추가될 수 있습니다.

**Comment by le2sky** [comment]
협력하는 객체의 내부 구조에 대한 결합으로 인해 발생하는 설계 문제를 해결하기 위해서 제안됐습니다. 디미터의 법칙을 따르기 위해서는 클래스가 특정한 조건을 만족하는 대상에게만 메시지를 전송하도록 해야합니다.

모든 클래스 C와 C에 구현된 메서드 M에 대해서, M이 메시지를 전송할 수 있는 모든 객체는 다음에 서술된 클래스의 인스턴스여야 합니다.

- M의 인자로 전달된 클래스(C 자체를 포함)
- C의 인스턴스 변수의 클래스

이때, M에 의해 생성된 객체나 M이 호출하는 메서드에 의해 생성된 객체, 전역 변수로 선언된 객체는 모두 M의 인자로 간주합니다. 이는 다음과 같이 요약할 수 있습니다.

낯선 자에게 말하지 말라.
오직 인접한 이웃하고만 말하라.
오직 하나의 도트만 사용하라.

디미터의 법칙은 객체가 자기 자신을 책임지는 자율적인 존재여야 한다는 사실을 강조합니다. 
정보를 처리하는 데 필요한 책임을 정보를 알고 있는 객체에게 할당하기 때문에 응집도가 높은 객체를 만들 수 있습니다. 
또한, 객체의 결합도를 낮추고,유지보수하기 쉬운 코드를 작성할 수 있습니다.

반면, 작성하는 코드량이 증가하는 것이 단점입니다.

**Comment by GIVEN53** [comment]
객체는 스스로를 책임지는 자율적인 존재
외부에서 객체 내부 접근을 최소화
